primo2/inference/dynamic.py

```diff
@@ -0,0 +1,305 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# This file is part of PRIMO2 -- Probabilistic Inference Modules.
+# Copyright (C) 2013-2017 Social Cognitive Systems Group, 
+#                         Faculty of Technology, Bielefeld University
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the Lesser GNU General Public License as 
+# published by the Free Software Foundation, either version 3 of the 
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public 
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+
+import abc
+import copy
+import six
+
+from . import exact
+
+
+__all__ = [
+    'PriorFeedbackExact',
+    'SoftEvidenceExact'
+]
+
+
+@six.add_metaclass(abc.ABCMeta)
+class DBNInferenceMethod(object):
+    """Abstract class for dynamic Bayesian network inference methods.
+
+    Contains the inference algorithm and the current state of the 
+    network at time t.
+    """
+
+    def __init__(self, dbn):
+        """Create the inference method and its state.
+
+        Parameters
+        ----------
+        dbn : DynamicBayesianNetwork
+            A 2-TBN dynamic Bayesian network structure.
+        """
+        super(DBNInferenceMethod, self).__init__()
+        self._dbn = dbn
+        self._t = 0
+        self._evidence = False
+
+    @abc.abstractmethod
+    def marginal_probabilities(self, variable):
```

Comment:
I don't think I'd put it in the Factor class as is, since a factor does not need to contain a valid probability. We would either provide a function which can quite likely return useless data, or we would need to extend the Factor class to be able to decide for itself if it contains a valid probability. If it currently represents a "joint probability" we could normalize it to sum to 1 in order to get the marginals but in the case of conditional probabilities we cannot simply do that.